### PR TITLE
re-write ec2_vpc_route_table.py to be Boto3 compatible [WIP]

### PR DESCRIPTION
**Pull request by linuxdynasty**
##### Issue Type:

- Feature Pull Request
##### Plugin Name:

ec2_vpc_route_table.py
##### Summary:

In the process of writing the ec2_vpc_nat_gateway.py, the ec2_vpc_route_table.py module would fail on subsequent runs. After further digging, Boto2 does not officially support Nat Gateways. This re-write of the module rips out Boto2 completely in place of Boto3. This module  officially supports all Amazon supported gateways. Also each function is heavily documented as well.
##### Example:

```
       [
            {
                "success": true,
                "changed": true,
                "msg": "Route table rtb-1234567 updated.",
                "associations": [
                    {
                        "subnet_id": "subnet-12345667",
                        "route_table_id": "rtb-1234567",
                        "main": false,
                        "route_table_association_id": "rtbassoc-1234567"
                    },
                    {
                        "subnet_id": "subnet-78654321",
                        "route_table_id": "rtb-78654321",
                        "main": false,
                        "route_table_association_id": "rtbassoc-78654321"
                    }
                ],
                "tags": [
                    {
                        "key": "Name",
                        "value": "dev_route_table"
                    },
                    {
                        "key": "env",
                        "value": "development"
                    }
                ],
                "routes": [
                    {
                        "gateway_id": "local",
                        "origin": "CreateRouteTable",
                        "state": "active",
                        "destination_cidr_block": "10.100.0.0/16"
                    },
                    {
                        "origin": "CreateRoute",
                        "state": "active",
                        "nat_gateway_id": "nat-12345678",
                        "destination_cidr_block": "0.0.0.0/0"
                    }
                ],
                "route_table_id": "rtb-1234567",
                "vpc_id": "vpc-1234567",
                "propagating_vgws": []
            }
        ]
```
- Rewrote the module to be Boto3 compatible.
- Documented every function.
- Supports all gateways (gateway_id, instance_id, network_interface_id, vpc_peering_connection_id, nat_gateway_id).

---

This change is [Reviewable](https://reviewable.io/reviews/ansible/ansible-modules-extras/1846)

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @linuxdynasty. To the current maintainers, @erydo, @wimnat please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by wimnat** [comment]
Unfortunately, i'm not sure what the core team will say about this. It breaks compatibility obviously.  @gregdek can you comment the core/aws guys here for input?

**Comment by linuxdynasty** [comment]
@wimnat is it what the module is returning, that breaks compatibility? Please let me know what aspect breaks it. I made sure that the module takes in the same parameters and in the same format.

**Comment by wimnat** [comment]
I mean that if someone didn't have boto3 installed on their system, it wouldn't run where as it could of been running fine before.

**Comment by linuxdynasty** [comment]
How can this issue be resolved? I know I can make most of the api calls be Boto compliant. What I am concerned with is, that if we say this module works with Nat Gateways, and the user who is testing this module doesn't have Boto3 installed, then this module will still not work. Do I add an exception just for nat gateways, to try and import boto3 and only use the boto3 api for that particular call? Or do we just not support the adding of nat gateways in a route?

I just searched through the ansible-extra-modules/cloud/amazon for boto3 and I do not see in the code where both boto and boto3 is being used. I see that boto is being imported, but I do not see modules that are actually performing a boto api call when boto3 is not available and vice versa.

Also only new features like Nat Gateway are being added to Boto3 and not Boto. Does that mean, any existing modules can not support newer features, due to the chicken and the egg problem?

Please let me know on how to proceed.

**Comment by linuxdynasty** [comment]
@sivel or @bcoca, may we get some guidance on how you guys think it is best to proceed with this PR?

**Comment by linuxdynasty** [comment]
@gregdek, @sivel , @bcoca , how should we continue with this PR? Should I close it and we do not support Amazon Nat Gateways?

**Comment by gregdek** [comment]
Moving to core review. I think that's most appropriate for now (since that's basically what @wimnat asked for.)

ping @ansible/core for input.

**Comment by gregdek** [comment]
Thanks @linuxdynasty for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by linuxdynasty** [comment]
@gregdek, I have rebased against development.

**Comment by gregdek** [comment]
Thanks @linuxdynasty. Putting back into core review.

**Comment by gregdek** [comment]
Thanks @linuxdynasty for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@linuxdynasty A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by linuxdynasty** [comment]
I am in the process of writing tests for this module.

**Comment by sgzijl** [comment]
@linuxdynasty thanks for the PR. According to the module documentation, it should be possible to work with subnet names, but when providing a subnet name the module will fail with the following error msg:

`An error occurred (InvalidSubnetID.NotFound) when calling the AssociateRouteTable operation: The subnet ID 'subnet-a' does not exist`

If I run into other issues, I will report them shortly.

**Comment by linuxdynasty** [comment]
@sgzijl thank you for catching that and I will update the tests accordingly.

**Comment by gregdek** [comment]
@linuxdynasty Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ryansb** [comment]
Is this rewrite going to allow legacy users to continue using boto with this module?

**Comment by linuxdynasty** [comment]
@ryansb this rewrite will not. This is Boto3 only. If that is an issue, I can close this PR out.

**Comment by wimnat** [comment]
I have brought this up before. At the Ansible 2.0 release, the core team specifically said they wanted to maintain compatibility with older boto modules but all new modules should be boto3.

Unless the core team have changed there stance then this module is cool, but perhaps not yet viable.

**Comment by ryansb** [comment]
@linuxdynasty after talking this over with the core team, we wanted new modules after 2.0 to be in boto3, but weren't going to _rewrite_ modules with boto3 until Ansible v2.3, which is in fact right around the corner. So keep this PR open, and once we start accepting 2.3 changes (in around a month) we'll be able to merge this change.

**Comment by ryansb** [comment]
Also, this needs to be rebased in any event, so in a few weeks do the rebase and re-ping. Thanks! :smile:

**Comment by gregdek** [comment]
@linuxdynasty A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@linuxdynasty Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@linuxdynasty A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Too hasty with that close... :) pinging @linuxdynasty again and resetting pending_action (since it was labeled WIP and bot never ran again.)

**Comment by gregdek** [comment]
Thanks @linuxdynasty for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@linuxdynasty A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@linuxdynasty Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2_vpc_route_table.py [hunk 1]
```diff
@@ -19,18 +19,21 @@
 short_description: Manage route tables for AWS virtual private clouds
 description:
     - Manage route tables for AWS virtual private clouds
-version_added: "2.0"
-author: Robert Estelle (@erydo), Rob White (@wimnat)
+version_added: "2.2"
```
Comment (ryansb) [inline cloud/amazon/ec2_vpc_route_table.py:22]: I don't think this should be changed, since this isn't new in the 2.2/2.3 release.

**Review by ryansb — CHANGES_REQUESTED** [review]

**Comment by wimnat** [comment]
@ryansb has the core team said we are allowed to rewrite modules to boto3 now we've hit 2.3?

File: cloud/amazon/ec2_vpc_route_table.py (hunk 1 above)
Comment (linuxdynasty) [reply cloud/amazon/ec2_vpc_route_table.py:22]: The reason I removed the other authors, was this was in fact a rewrite of the original module.

Comment (linuxdynasty) [reply cloud/amazon/ec2_vpc_route_table.py:22]: I have no issues, re-adding them though.

**Comment by ryansb** [comment]
@wimnat yes, that's correct. 

@linuxdynasty can you migrate this PR to the now-main `ansible/ansible` repo using [prmover](https://prmover.pythonanywhere.com/)?

**Comment by linuxdynasty** [comment]
Sure thing AND YEAHHHHH BUDDDYYYYY BOTO3!!!!!

**Pull request closed without merging**